File: cmd/07_autoencoder_compute_graph/main.go
package main

import (
	"encoding/csv"
	"github.com/Jimmy2099/torch/data_store/compute_graph"
	"github.com/Jimmy2099/torch/data_store/network"
	"github.com/Jimmy2099/torch/data_store/tensor"
	"github.com/Jimmy2099/torch/pkg/fmt"
	"github.com/Jimmy2099/torch/pkg/log"
	"github.com/Jimmy2099/torch/testing"
	"io"
	"math/rand"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

type AutoEncoder struct {
	graph *compute_graph.ComputationalGraph
}

func transpose2D(data []float32, rows, cols int) ([]float32, []int) {
	transposed := make([]float32, rows*cols)
	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			transposed[j*rows+i] = data[i*cols+j]
		}
	}
	return transposed, []int{cols, rows}
}

func loadParameters() (map[string]*tensor.Tensor, map[string]*tensor.Tensor) {
	weights := make(map[string]*tensor.Tensor)
	biases := make(map[string]*tensor.Tensor)

	layerSpecs := []struct {
		name        string
		weightShape []int
		biasShape   []int
	}{
		{"encoder.0", []int{128, 784}, []int{128}},
		{"encoder.2", []int{64, 128}, []int{64}},
		{"encoder.4", []int{32, 64}, []int{32}},
		{"decoder.0", []int{64, 32}, []int{64}},
		{"decoder.2", []int{128, 64}, []int{128}},
		{"decoder.4", []int{784, 128}, []int{784}},
	}

	basePath, err := filepath.Abs("./py/data")
	if err != nil {
		panic(err)
	}

	fmt.Printf("Loading parameters from: %s\n", basePath)

	for _, spec := range layerSpecs {
		weightPath := filepath.Join(basePath, spec.name+".weight.csv")
		weightTensor, err := loadTensorFromCSV(weightPath, spec.weightShape)
		if err != nil {
			panic(fmt.Sprintf("Load weight %s failed: %v", weightPath, err))
		}
		transposedData, transposedShape := transpose2D(weightTensor.Data, spec.weightShape[0], spec.weightShape[1])
		weights[spec.name] = tensor.NewTensor(transposedData, transposedShape)

		biasPath := filepath.Join(basePath, spec.name+".bias.csv")
		biasTensor, err := loadTensorFromCSV(biasPath, spec.biasShape)
		if err != nil {
			panic(fmt.Sprintf("Load bias %s failed: %v", biasPath, err))
		}
		biases[spec.name] = biasTensor
	}
	return weights, biases
}

func NewAutoEncoder() *AutoEncoder {
	fmt.Println("Initializing AutoEncoder with computational graph...")
	graph := compute_graph.NewComputationalGraph()
	net := graph.Network
	nodes := make(map[string]*network.Node)

	createNode := func(name, nodeType string) *network.Node {
		node := net.NewNode()
		node.Name = name
		node.Type = nodeType
		nodes[name] = node
		return node
	}

	createNode("input", "Tensor_Input")
	paramNames := []string{"encoder.0", "encoder.2", "encoder.4", "decoder.0", "decoder.2", "decoder.4"}
	for _, p := range paramNames {
		createNode(p+".weight", "Tensor_Weight")
		createNode(p+".bias", "Tensor_Bias")
	}

	fmt.Println("Building computation graph blueprint...")
	currentX := "input"

	buildLayer := func(inputName, layerName, activation string) string {
		matmulNode := createNode(layerName+"_matmul", "MatMul")
		matmulOut := createNode(layerName+"_matmul_out", "Tensor_Hidden")
		matmulNode.AddInput(nodes[inputName])
		matmulNode.AddInput(nodes[layerName+".weight"])
		matmulNode.AddOutput(matmulOut)

		addNode := createNode(layerName+"_add", "Add")
		addOut := createNode(layerName+"_add_out", "Tensor_Hidden")
		addNode.AddInput(matmulOut)
		addNode.AddInput(nodes[layerName+".bias"])
		addNode.AddOutput(addOut)

		if activation != "" {
			actNode := createNode(layerName+"_"+activation, activation)
			actOut := createNode(layerName+"_"+activation+"_out", "Tensor_Hidden")
			actNode.AddInput(addOut)
			actNode.AddOutput(actOut)
			return actOut.Name
		}
		return addOut.Name
	}

	currentX = buildLayer(currentX, "encoder.0", "Relu")
	currentX = buildLayer(currentX, "encoder.2", "Relu")
	currentX = buildLayer(currentX, "encoder.4", "")

	currentX = buildLayer(currentX, "decoder.0", "Relu")
	currentX = buildLayer(currentX, "decoder.2", "Relu")
	currentX = buildLayer(currentX, "decoder.4", "Sigmoid")

	net.AddInput(nodes["input"])
	for _, p := range paramNames {
		net.AddInput(nodes[p+".weight"])
		net.AddInput(nodes[p+".bias"])
	}
	net.AddOutput(nodes[currentX])

	fmt.Println("\nLoading and setting parameters...")
	weights, biases := loadParameters()
	for name, tensorData := range weights {
		graph.GetTensorByName(name + ".weight").SetValue(tensorData)
	}
	for name, tensorData := range biases {
		graph.GetTensorByName(name + ".bias").SetValue(tensorData)
	}

	ae := &AutoEncoder{
		graph: graph,
	}

	fmt.Println("Computation graph structure:")
	return ae
}

func (ae *AutoEncoder) Forward(x *tensor.Tensor) *tensor.Tensor {
	if x.Size() != 784 {
		panic("Invalid input size, must be 784")
	}
	x.Reshape([]int{1, 784})

	ae.graph.GetTensorByName("input").SetValue(x)

	ae.graph.Forward()

	outputTensor := ae.graph.GetTensorByName(ae.graph.Network.GetOutputName()[0])
	return outputTensor.Value()
}

func loadTensorFromCSV(path string, shape []int) (*tensor.Tensor, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	reader := csv.NewReader(file)
	records, err := reader.ReadAll()
	if err != nil {
		return nil, err
	}

	data := make([]float32, 0, len(records)*len(records[0]))
	for _, row := range records {
		for _, val := range row {
			f, err := strconv.ParseFloat(val, 32)
			if err != nil {
				return nil, err
			}
			data = append(data, float32(f))
		}
	}

	return tensor.NewTensor(data, shape), nil
}

func main() {
	{
		directory := "./py/mnist_noisy_images"

		if _, err := os.Stat(directory); err != nil {
			d, _ := os.Getwd()
			runCommand(filepath.Join(d, "py"), filepath.Join(d, "py", "generate_go_testdata.py"))
		}

		images, labels, err := LoadDataFromCSVDir(directory)
		if err != nil {
			log.Fatalf("Error loading data: %v", err)
		}

		if len(images) > 0 && len(labels) > 0 {
			fmt.Printf("First Image Data:\n%v\n", images[0])
			fmt.Printf("First Image Label: %s\n", labels[0])
		}

		model := NewAutoEncoder()

		{
			model.graph.ComputeDependencyGraph.ComputeSortedNodes()
			err = model.graph.Validate()
			if err != nil {
				panic(err)
			}
		}

		for num := 0; num < len(images); num++ {
			prediction := Predict(model, images[num])
			prediction.SaveToCSV(filepath.Join(directory, strings.Replace(labels[num], ".png.csv", ".png.denoise.csv", -1)))
		}

		{
			var onnx *compute_graph.ONNX
			onnx, err = model.graph.ToONNXModel()
			if err != nil {
				log.Fatalf("Error creating ONNX model: %v", err)
			}
			err = onnx.SaveONNX("model.onnx")
			if err != nil {
				log.Fatalf("Error saving ONNX model: %v", err)
			}
		}

		fmt.Println("label:", labels)
		var fileName []string
		{

			for i := 0; i < len(labels); i++ {
				d, _ := os.Getwd()
				name := filepath.Join(d, "py", "mnist_noisy_images", labels[i])
				name = strings.Replace(name, ".csv", "", -1)
				fileName = append(fileName, name)
				fmt.Println()
			}
			if !testing.IsTesting() {
				predictPlot(fileName)
			}
		}

	}
}

func ReadCSV(filepath string) (*tensor.Tensor, error) {
	file, err := os.Open(filepath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	reader := csv.NewReader(file)
	lines, err := reader.ReadAll()
	if err != nil {
		return nil, err
	}

	rows := len(lines)
	cols := len(lines[0])
	if rows*cols != 784 {
		return nil, fmt.Errorf("expected 784 values in CSV, got %d", rows*cols)
	}

	data := make([]float32, rows*cols)

	for i, line := range lines {
		for j, value := range line {
			val, err := strconv.ParseFloat(value, 64)
			if err != nil {
				return nil, err
			}
			data[i*cols+j] = float32(val)
		}
	}

	tensorImage := tensor.NewTensor(data, []int{28, 28})

	return tensorImage, nil
}

func LoadDataFromCSVDir(directory string) ([]*tensor.Tensor, []string, error) {
	var tensors []*tensor.Tensor
	var labels []string

	labelFilePath := filepath.Join(directory, "labels.csv")
	labelFile, err := os.Open(labelFilePath)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to open label file: %v", err)
	}
	defer labelFile.Close()

	reader := csv.NewReader(labelFile)
	labelRecords, err := reader.ReadAll()
	if err != nil {
		return nil, nil, fmt.Errorf("failed to read label CSV: %v", err)
	}

	labelMap := make(map[string]string)
	for _, record := range labelRecords {
		if len(record) >= 2 {
			filename := strings.TrimSpace(record[0])
			label := strings.TrimSpace(record[1])
			labelMap[filename] = label
		}
	}

	files, err := os.ReadDir(directory)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to read directory: %v", err)
	}

	for _, file := range files {
		if file.IsDir() || file.Name() == "labels.csv" {
			continue
		}
		if !strings.HasSuffix(file.Name(), "png.csv") {
			continue
		}

		imagePath := filepath.Join(directory, file.Name())
		image, err := ReadCSV(imagePath)
		if err != nil {
			log.Printf("failed to read image file %s: %v", file.Name(), err)
			continue
		}

		label := file.Name()

		tensors = append(tensors, image)
		labels = append(labels, label)
	}

	return tensors, labels, nil
}

func Predict(model *AutoEncoder, image *tensor.Tensor) *tensor.Tensor {
	image = image.Reshape([]int{1, 1, 28, 28})
	output := model.Forward(image)

	output.Reshape([]int{28, 28})
	return output
}

func predictPlot(imagePaths []string) error {

	rand.Seed(time.Now().UnixNano())
	suffix := fmt.Sprintf("%d", rand.Intn(1000000))

	tmpScriptFileName := filepath.Join(os.TempDir(), "predict_plot_"+suffix+".py")
	tmpDataFileName := filepath.Join(os.TempDir(), "image_predictions_"+suffix+".txt")

	pythonScript := `
import sys
import matplotlib.pyplot as plt
import matplotlib.image as mpimg
import os
import numpy as np

def load_numpy_from_csv(file_path):
    with open(file_path, 'r') as f:
        header = f.readline().strip()
        if not header.startswith("shape,"):
            raise ValueError("Invalid CSV format: missing shape header")
        
        shape = list(map(int, header.split(",")[1:]))
        data = np.loadtxt(f, delimiter=",")
    
    flattened = data.flatten()
    return (flattened).reshape(*shape)


def load_data(file_path):
    image_paths = []
    predictions = []
    with open(file_path, 'r') as f:
        for line in f:
            parts = line.strip().split(',')
            image_paths.append(parts[0])
            predictions.append(parts[1])
    return image_paths, predictions

def predict_plot(data_file):
    image_paths, predictions = load_data(data_file)
    num = len(image_paths)
    fig, axes = plt.subplots(2, num, figsize=(15, 5)) 

    for i, (img_path, denoise_csv_path) in enumerate(zip(image_paths, predictions)):
        img = mpimg.imread(img_path)
        axes[0,i].imshow(img, cmap='gray')
        print(denoise_csv_path)
        print(f"Attempting to load: {denoise_csv_path}")
        image_data = load_numpy_from_csv(denoise_csv_path)
        axes[1,i].imshow(image_data, cmap='gray')
        os.remove(denoise_csv_path)
        axes[0,i].axis('off')
        axes[1,i].axis('off')

    plt.tight_layout()
    plt.savefig("predictions.png")
    plt.show()

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: python predict_plot.py <data_file_path>")
        sys.exit(1)
    data_file = sys.argv[1]
    predict_plot(data_file)
`
	scriptFile, err := os.Create(tmpScriptFileName)
	if err != nil {
		return fmt.Errorf("unable to create temp Python script: %v", err)
	}
	_, err = scriptFile.WriteString(pythonScript)
	if err != nil {
		scriptFile.Close()
		return fmt.Errorf("unable to write Python script: %v", err)
	}
	scriptFile.Close()
	defer os.Remove(tmpScriptFileName)

	dataFile, err := os.Create(tmpDataFileName)
	if err != nil {
		return fmt.Errorf("unable to create temp data file: %v", err)
	}
	for i := 0; i < len(imagePaths); i++ {
		_, err = dataFile.WriteString(fmt.Sprintf("%s,%s\n", imagePaths[i], imagePaths[i]+".denoise.csv"))
		if err != nil {
			dataFile.Close()
			return fmt.Errorf("unable to write to temp data file: %v", err)
		}
	}
	dataFile.Close()
	defer os.Remove(tmpDataFileName)

	cmd := exec.Command("python", tmpScriptFileName, tmpDataFileName)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr

	if err = cmd.Run(); err != nil {
		return fmt.Errorf("error executing Python script: %v", err)
	}

	return nil
}

func runCommand(workSpace string, fileName string) {

	cmd := exec.Command("python", fileName)
	cmd.Dir = workSpace
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		panic(fmt.Sprintln("Error creating Stdout pipe:", err))
	}

	stderr, err := cmd.StderrPipe()
	if err != nil {
		panic(fmt.Sprintln("Error creating Stderr pipe:", err))
	}

	if err = cmd.Start(); err != nil {
		panic(fmt.Sprintln("Error starting Python script:", err))
	}

	go func() {
		_, err = io.Copy(os.Stdout, stdout)
		if err != nil {
			panic(fmt.Sprintln("Error copying stdout:", err))
		}
	}()

	go func() {
		_, err = io.Copy(os.Stderr, stderr)
		if err != nil {
			panic(fmt.Sprintln("Error copying stderr:", err))
		}
	}()

	if err = cmd.Wait(); err != nil {
		panic(fmt.Sprintln("Error waiting for Python script to finish:", err))
	}
}

File: cmd/07_autoencoder_compute_graph/main.py
import onnxruntime as ort
import numpy as np
import os
import csv
import glob
import tempfile
import subprocess
import random

class AutoEncoderONNX:
    def __init__(self, model_path, data_dir="py/data"):
        if not os.path.exists(model_path):
            raise FileNotFoundError(f"Model file not found at {model_path}")

        self.session = ort.InferenceSession(model_path)

        self.input_names = [input.name for input in self.session.get_inputs()]
        self.output_name = self.session.get_outputs()[0].name

        print(f"Model loaded successfully. Input names: {self.input_names}")
        print(f"Output name: {self.output_name}")

        self.parameters = self.load_parameters(data_dir)

        for param_name, param_data in self.parameters.items():
            print(f"Parameter {param_name}: shape {param_data.shape}")

    def load_parameters(self, data_dir):
        parameters = {}

        layer_specs = [
            ("encoder.0", [128, 784], [128]),
            ("encoder.2", [64, 128], [64]),
            ("encoder.4", [32, 64], [32]),
            ("decoder.0", [64, 32], [64]),
            ("decoder.2", [128, 64], [128]),
            ("decoder.4", [784, 128], [784]),
        ]

        for name, weight_shape, bias_shape in layer_specs:
            weight_path = os.path.join(data_dir, f"{name}.weight.csv")
            if os.path.exists(weight_path):
                weight_data = self.load_tensor_from_csv(weight_path, weight_shape)
                weight_data = weight_data.reshape(weight_shape).T
                parameters[f"{name}.weight"] = weight_data.astype(np.float32)
                print(f"Loaded weight {name}: {weight_data.shape}")
            else:
                print(f"Warning: Weight file not found: {weight_path}")

            bias_path = os.path.join(data_dir, f"{name}.bias.csv")
            if os.path.exists(bias_path):
                bias_data = self.load_tensor_from_csv(bias_path, bias_shape)
                parameters[f"{name}.bias"] = bias_data.astype(np.float32)
                print(f"Loaded bias {name}: {bias_data.shape}")
            else:
                print(f"Warning: Bias file not found: {bias_path}")

        return parameters

    def load_tensor_from_csv(self, path, shape):
        data = []
        with open(path, 'r', encoding='utf-8') as f:
            reader = csv.reader(f)
            for row in reader:
                data.extend([float(x) for x in row])

        return np.array(data).reshape(shape)

    def preprocess(self, image_data):
        data = np.array(image_data, dtype=np.float32)

        if data.size == 784:
            data = data.reshape(1, 784)
        else:
            raise ValueError("Input data size must be 784 (28x28)")
        return data

    def predict(self, input_data):
        processed_data = self.preprocess(input_data)
        print(f"Input data shape: {processed_data.shape}")

        input_feed = {"input": processed_data}

        for param_name, param_data in self.parameters.items():
            input_feed[param_name] = param_data

        try:
            outputs = self.session.run(
                [self.output_name],
                input_feed
            )
            print(f"Output shape: {outputs[0].shape}")
            return outputs[0]
        except Exception as e:
            print(f"Error during inference: {e}")
            raise

    def save_to_csv(self, data, file_path):
        reshaped = data.reshape(28, 28)

        os.makedirs(os.path.dirname(file_path), exist_ok=True)

        with open(file_path, 'w', newline='', encoding='utf-8') as f:
            writer = csv.writer(f)
            writer.writerow(["shape"] + list(reshaped.shape))
            for row in reshaped:
                writer.writerow(row)
        print(f"Saved denoised image to {file_path}")

def load_image_from_csv(file_path):
    data = []
    with open(file_path, 'r', encoding='utf-8') as f:
        reader = csv.reader(f)

        first_row = next(reader, None)
        if first_row and first_row[0] == "shape":
            for row in reader:
                data.extend([float(x) for x in row])
        else:
            if first_row:
                data.extend([float(x) for x in first_row])
            for row in reader:
                data.extend([float(x) for x in row])

    if len(data) != 784:
        raise ValueError(f"Expected 784 values in CSV, got {len(data)}")

    return np.array(data)

def load_data_from_csv_dir(directory):
    images = []
    labels = []

    label_file = os.path.join(directory, "labels.csv")
    label_map = {}

    if os.path.exists(label_file):
        with open(label_file, 'r', encoding='utf-8') as f:
            reader = csv.reader(f)
            for row in reader:
                if len(row) >= 2:
                    filename = row[0].strip()
                    label = row[1].strip()
                    label_map[filename] = label

    csv_files = glob.glob(os.path.join(directory, "*.png.csv"))
    if not csv_files:
        csv_files = glob.glob(os.path.join(directory, "*.csv"))
        csv_files = [f for f in csv_files if not f.endswith("labels.csv")]

    for file_path in csv_files:
        try:
            image_data = load_image_from_csv(file_path)
            filename = os.path.basename(file_path)

            images.append(image_data)
            labels.append(filename)
            print(f"Successfully loaded {filename}")
        except Exception as e:
            print(f"Error loading {file_path}: {str(e)}")

    return images, labels

def predict_plot(image_paths):
    suffix = str(random.randint(0, 1000000))
    tmp_script = os.path.join(tempfile.gettempdir(), f"predict_plot_{suffix}.py")
    tmp_data = os.path.join(tempfile.gettempdir(), f"image_predictions_{suffix}.txt")

    python_script = '''
import sys
import matplotlib.pyplot as plt
import matplotlib.image as mpimg
import os
import numpy as np

def load_numpy_from_csv(file_path):
    with open(file_path, 'r', encoding='utf-8') as f:
        header = f.readline().strip()
        if not header.startswith("shape,"):
            # Fallback for old format
            f.seek(0)
            data = np.loadtxt(f, delimiter=",")
            return data

        shape_str = header.split(",")[1:]
        if not all(s.isdigit() for s in shape_str):
             raise ValueError("Invalid shape format in CSV header")

        shape = tuple(map(int, shape_str))
        data = np.loadtxt(f, delimiter=",")
    
    return data.reshape(shape)

def load_data(file_path):
    image_paths = []
    predictions = []
    with open(file_path, 'r', encoding='utf-8') as f:
        for line in f:
            parts = line.strip().split(',')
            if len(parts) == 2:
                image_paths.append(parts[0])
                predictions.append(parts[1])
    return image_paths, predictions

def predict_plot(data_file):
    image_paths, predictions = load_data(data_file)
    num = len(image_paths)
    if num == 0:
        print("No valid image-prediction pairs found to plot.")
        return
        
    fig, axes = plt.subplots(2, num, figsize=(15, 5)) 

    # Handle case where there is only one image
    if num == 1:
        axes = np.array(axes).reshape(2, 1)

    for i, (img_path, denoise_csv_path) in enumerate(zip(image_paths, predictions)):
        if not os.path.exists(img_path):
            print(f"Warning: Image file not found: {img_path}")
            axes[0,i].text(0.5, 0.5, 'Original not found', ha='center', va='center')
            axes[0,i].axis('off')
        else:
            try:
                img = mpimg.imread(img_path)
                axes[0,i].imshow(img, cmap='gray')
                axes[0,i].set_title("Original")
            except Exception as e:
                print(f"Error reading image {img_path}: {e}")
                axes[0,i].text(0.5, 0.5, 'Error loading image', ha='center', va='center')

        if not os.path.exists(denoise_csv_path):
            print(f"Warning: Denoised CSV file not found: {denoise_csv_path}")
            axes[1,i].text(0.5, 0.5, 'Denoised not found', ha='center', va='center')
            axes[1,i].axis('off')
        else:
            try:
                image_data = load_numpy_from_csv(denoise_csv_path)
                axes[1,i].imshow(image_data, cmap='gray')
                axes[1,i].set_title("Denoised")
            except Exception as e:
                print(f"Error loading denoised CSV {denoise_csv_path}: {e}")
                axes[1,i].text(0.5, 0.5, 'Error loading CSV', ha='center', va='center')

        axes[0,i].axis('off')
        axes[1,i].axis('off')

    plt.tight_layout()
    plt.savefig("predictions.png")
    plt.show()

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: python predict_plot.py <data_file_path>")
        sys.exit(1)
    data_file = sys.argv[1]
    predict_plot(data_file)
'''

    with open(tmp_script, 'w', encoding='utf-8') as f:
        f.write(python_script)

    with open(tmp_data, 'w', encoding='utf-8') as f:
        for img_path in image_paths:
            if not os.path.exists(img_path):
                print(f"Warning: Original image not found: {img_path}")
                continue

            denoise_filename = os.path.basename(img_path).replace(".png", ".png.denoise.csv")
            denoise_path = os.path.join("results", denoise_filename)

            if not os.path.exists(denoise_path):
                print(f"Warning: Denoised file not found: {denoise_path}")
                continue

            f.write(f"{img_path},{denoise_path}\n")

    try:
        result = subprocess.run(['python', tmp_script, tmp_data],
                                capture_output=True, text=True, check=False)
        print("Python script output:")
        print(result.stdout)
        if result.stderr:
            print("Python script errors:")
            print(result.stderr)

        results_dir = "results"
        os.makedirs(results_dir, exist_ok=True)
        if os.path.exists("predictions.png"):
            # 解决windows下rename时目标文件已存在的问题
            target_path = os.path.join(results_dir, "predictions.png")
            if os.path.exists(target_path):
                os.remove(target_path)
            os.rename("predictions.png", target_path)
            print(f"Plot saved to {target_path}")

    finally:
        if os.path.exists(tmp_script):
            os.remove(tmp_script)
        if os.path.exists(tmp_data):
            os.remove(tmp_data)

if __name__ == "__main__":
    MODEL_PATH = "ae_model.onnx"
    DATA_DIR = "py/mnist_noisy_images"
    PARAM_DIR = "py/data"
    RESULTS_DIR = "results"

    try:
        if not os.path.exists(PARAM_DIR):
            print(f"Parameter directory {PARAM_DIR} not found. Please check the path.")
            exit(1)

        autoencoder = AutoEncoderONNX(MODEL_PATH, PARAM_DIR)

        images, labels = load_data_from_csv_dir(DATA_DIR)
        print(f"Loaded {len(images)} images from {DATA_DIR}")

        if len(images) == 0:
            print("No images found. Please check the data directory.")
            exit(1)

        os.makedirs(RESULTS_DIR, exist_ok=True)

        image_paths = []
        for i, (image_data, label) in enumerate(zip(images, labels)):
            print(f"Processing image {i+1}/{len(images)}: {label}")

            denoised = autoencoder.predict(image_data)

            if label.endswith(".png.csv"):
                output_filename = label.replace(".png.csv", ".png.denoise.csv")
            else:
                output_filename = label.replace(".csv", ".denoise.csv")

            output_path = os.path.join(RESULTS_DIR, output_filename)
            autoencoder.save_to_csv(denoised[0], output_path)

            if label.endswith(".png.csv"):
                original_image_path = os.path.join(DATA_DIR, label.replace(".png.csv", ".png"))
            else:
                original_image_path = os.path.join(DATA_DIR, label.replace(".csv", ".png"))

            if os.path.exists(original_image_path):
                image_paths.append(original_image_path)
            else:
                print(f"Warning: Original image not found: {original_image_path}")

        if image_paths:
            print("Generating plot...")
            predict_plot(image_paths)
            print("Predictions saved and plotted successfully")
        else:
            print("No original images found to generate a plot.")

    except Exception as e:
        print(f"Error: {str(e)}")
        import traceback
        traceback.print_exc()